Split bugfix tst

markuspf opened the pull request:
This PR splits bugfix.tst into separate testfiles. This is work-in-progress, so do not merge yet, I just want to test this on travis as I go along.

fingolfin commented:
What exactly does this gain us? If you just want to know which tests ate slow, then the `showProgress` option for `Test` would also do that.

ChrisJefferson commented:
I don't know what the right amount of splitting is, but some amount of splitting is useful, when this test fails I find it hard to (a) figure out exactly where it failed, and (b) how much of it I have to cut out and run seperately, when I'm trying to debug.

I would be happy with it just being split up a little.. perhaps yearly?

fingolfin commented on tst/testbugfix/00030.tst:
```diff
@@ -0,0 +1,20 @@
+  <two-sided ideal in <Lie algebra of dimension 6 over Rationals>, 
+      (dimension 3)> ]
+
+##  bug 16 for fix 5
```
So how exactly did you split the tests? With a script, or manually, or a mix?  What is the condition upon which you made a split -- e.g. why is there no split here?

markuspf replied:
Just for the record: I split `bugfix.tst` by using a script that basically looked for places where there was an empty line followed by a comment. It worked surprisingly well, of course far from perfectly well!

fingolfin commented on tst/testbugfix/00060.tst:
```diff
@@ -0,0 +1,53 @@
+  IdSmallGroup is not available for this size. 
+ 
+
+# 2005/05/04 (SL)
```
Here is another spot where it's unclar to me why there is no split here?

fingolfin commented on tst/testbugfix/00065.tst:
```diff
@@ -0,0 +1,5 @@
+gap> AClosestVectorCombinationsMatFFEVecFFECoords;
+<Operation "AClosestVectorCombinationsMatFFEVecFFECoords">
+gap> ConstituentsPolynomial;
+function( p ) ... end
```
I know you are not cleaning up these tests right now, but here is one wher I am really confused as to what it is supposed to test, exactly...

fingolfin commented on tst/testbugfix/00066.tst:
```diff
@@ -0,0 +1,3 @@
+# 2005/04/27 (TB)
+gap> IsBound( CycList );
+true
```
Another one where I am not sure what the point of the tests is?

fingolfin commented on tst/testbugfix/00078.tst:
```diff
@@ -0,0 +1,3 @@
+# 2005/07/13 (JS)
+gap> PerfectGroup(7800,1);; # load perf2.grp
+gap> PerfectGroup(7680,1);; # should load perf1.grp, gives error in 4.4.5
```
This tests is redundant now, we have a test that loads the whole perfect groups library.

fingolfin commented on tst/testbugfix/00068.tst:
```diff
@@ -0,0 +1,5 @@
+gap> NumberSmallGroups(5^6);
+684
+gap> NumberSmallGroups(5*7*9*11*13);
+22
```
I gues right now the tests are in the exact same order they appeared in in testbugfix.tst. But I wonder if we could try to automatically group them in a bunch of categries. E.g. if `SmallGroup` occurs, then it' likely a test for small groups. (Of course not always, since some tests will involve small groups just as input for another test). Same for `TransitiveGroup`, `PrimitiveGroup`, etc.

Anyway, that's a thought for later, I guess... I am assuming we will want to remove numbered tests which are subsumed by other tests now, and also group some of these tsts in a more logical manner...

fingolfin commented on tst/testbugfix/00069.tst:
```diff
@@ -0,0 +1,3 @@
+# 2005/05/05 (TB)
+gap> IsBound( ShowPackageVariables );
+true
```
Another test which just checks whether some identifier is defined. What is the point here? Preventing us from accidentally removing any documented identifiers? But then, shouldn't we rather have a full test for this, e.g. get a list of all documented identifiers, and verify they are defined?

fingolfin replied:
(Might be an idea for a a feature request issue (FRI) / request for enhancement (RFE)

alex-konovalov replied:
Looks like a relic from times when we were supplying bugfix archives, to unpack and overwrite existing files. Not sure we need this any more.

fingolfin commented on tst/testbugfix/00088.tst:
```diff
@@ -0,0 +1,5 @@
+gap> f:=function() Assert(0,false); end;; g:=function() f(); end;;
+gap> ##  The following should just trigger a normal error, but in 4.4.5
+gap> ##  it will send a few hundred lines before crashing:
+gap> # g();
```
This is a no-test, as `g()` is commented out... Shall we remove the comment? (again, it's OK by me if we do this after the PR is merged... I am just not sure what exactly the plan is, so I am adding comments as I look at this)

fingolfin commented on tst/testbugfix/00094.tst:
```diff
@@ -0,0 +1,6 @@
+# since a different message is printed on 32 bit systems and 64 bit systems
+gap> a := 2^(8*GAPInfo.BytesPerVariable-4)-1;;
+gap> Unbind( x );
+gap> # x := [-a..a];;
```
We now have directories for 32/64 bit specific tests, so I guess this should be copied to the two, with the last line activated

fingolfin commented on tst/testbugfix/00095.tst:
```diff
@@ -0,0 +1,3 @@
+# Range: the length of a range must be less than 2^28
+gap> IsBound(x);
+false
```
This belongs to the preceding test

fingolfin commented on tst/testbugfix/00098.tst:
```diff
@@ -0,0 +1,3 @@
+# 2005/08/26 (Max)
+gap> IsOperation(MutableCopyMat);
+true
```
So the tests I am musing about above (checking all documented words for being defined) could also check them to match what the documentation says: is it really an operation / property / attribute / function / ... ?

fingolfin commented on tst/testbugfix/00109.tst:
```diff
@@ -0,0 +1,22 @@
+  [ 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 5, 5, 17, 353, 1201, 169553, 7699649, 
+      134818753, 47072139617, 531968664833 ] ]
+
+# 2005/08/24 (SL, FL)
```
Split here?

fingolfin commented on tst/testbugfix/00112.tst:
```diff
@@ -0,0 +1,4 @@
+# 2005/09/20 (SK)
+gap> # None as the library methods for `NormalSubgroups' apparently obey
+gap> # the `rule' that the trivial subgroup appears in the first and the
+gap> # whole group appears in the last position.
```
That's not a test, just a comment... Huh?

fingolfin commented on tst/testbugfix/00242.tst:
```diff
@@ -0,0 +1,79 @@
+gap> r := RingByStructureConstants([7,5],T);
+<ring with 2 generators>
+gap> Characteristic(r);
+35
```
This could go into a new file `testinstall/opers/Characteristic.tst`

alex-konovalov commented:
@fingolfin wrote:
> All in all, I think now this is a good idea, as a step towards improving the test suite. After this split, we can git mv invidual tests files to have more indicative names, or to group them together into subdirs (under testbugfix or testinstall or ... as appropriate) etc.

Indeed. These tests were formerly collected from `dev/Updates` entries, where they had some meaningful names...

markuspf commented:
I split the files that showed up and removed the comment-only one. Maybe it was misguided `--force`ing it, but well. I suggest merging and then continuing the cleanup with a separate PR?

alex-konovalov commented:
- Have you had some sanity check - e.g. `cat` all files produced by your script and check that `diff` with the original file has only blank lines, but nothing else?

- What about trying some naming scheme? `YYYY-MM-DD-those-nice-things`?

markuspf commented:
The diffs between `cat *` and `bugfix.tst` are comments and what I have removed.

I don't know whether a naming scheme makes any sense really (nor would I know how to name the files right now).

alex-konovalov commented:
Of course consecutive numbers are also Ok, they are located in chronological order already... and the date may be in the test file in the comment... and for newer files it will be in revision history. It will be tedious to extract names of dev/update entries from which tests were extracted (and also not all of them ended up in the test file that way).

So, maybe the naming scheme should be `00XYZ-something-made-working.tst` and with 5 digits we will have some time to add more bugfix tests. The further `-something-made-working` should be short and informative to ease navigation, and with the number preceding it one should not care too much about uniqueness. For example, `00000-install-immediate-globals.tst`, `00001-irreducible-characters.tst` etc. But this can be done later, so I'm happy with merging this now.

ChrisJefferson commented:
One final question before merging, have you tried running each test in an independent GAP session? (to make sure no tests are relying on the output of others).

alex-konovalov commented:
This is btw what `make testupdates` does - it extracts tests from dev/Updates entries and indeed runs each in a separate GAP session. Should be adaptable to make this check.

markuspf commented:
@ChrisJefferson I now did, and found some problems that I'll address before merge

fingolfin commented on tst/test64bit/range.tst:
```diff
@@ -0,0 +1,9 @@
+# 2005/08/23 (FL)
+# commented out the test and the error message,
+# since a different message is printed on 32 bit systems and 64 bit systems
```
That comment is no longer relevant

fingolfin commented on tst/test32bit/range.tst:
```diff
@@ -0,0 +1,9 @@
+# 2005/08/23 (FL)
+# commented out the test and the error message,
+# since a different message is printed on 32 bit systems and 64 bit systems
```
That comment is no longer relevant
